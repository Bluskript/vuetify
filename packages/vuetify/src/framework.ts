import { inject } from 'vue'
import { VuetifyThemeSymbol, createTheme } from './composables'

// Types
import type { InjectionKey, App } from 'vue'

export interface VuetifyComponentDefaults {
  [key: string]: Record<string, unknown>
  global: Record<string, unknown>
}

export interface VuetifyInstance {
  defaults: VuetifyComponentDefaults
}

export interface VuetifyOptions {
  components?: Record<string, any>
  directives?: Record<string, any>
  defaults?: Partial<VuetifyComponentDefaults>
}

export const VuetifySymbol: InjectionKey<VuetifyInstance> = Symbol.for('vuetify')

export const useVuetify = () => {
  const vuetify = inject(VuetifySymbol)

  if (!vuetify) {
    throw new Error('Vuetify has not been installed on this app')
  }

  return vuetify
}

export const createVuetify = (options: VuetifyOptions = {}) => {
  const install = (app: App) => {
    const {
      components = {},
      directives = {},
      defaults = {},
    } = options

    for (const key in directives) {
      const directive = directives[key]

      app.directive(key, directive)
    }

    for (const key in components) {
      const component = components[key]

      app.component(key, component)
    }

    const vuetify: VuetifyInstance = {
      defaults: {
        global: {},
        ...defaults,
      },
    }

    app.provide(VuetifySymbol, vuetify)
    app.config.globalProperties.$vuetify = vuetify
    app.provide(VuetifyThemeSymbol, createTheme({
      defaultTheme: 'light',
      themes: {
        light: {
          background: '#ffffff',
          error: '#ff0000',
          warning: '#ffff00',
          info: '#0000ff',
          success: '#00ff00',
          primary: '#ff0000',
          primaryVariant: '#ff00ff',
          secondary: '#0000ff',
          secondaryVariant: '#4444ff',
          surface: '#ffffff',
          inverseText: '#ffffff',
          text: '#000000',
        },
        dark: {
          background: '#222222',
          error: '#ff0000',
          warning: '#ffff00',
          info: '#0000ff',
          success: '#00ff00',
          primary: '#ff00ff',
          primaryVariant: '#33ff33',
          secondary: '#555500',
          secondaryVariant: '#4444ff',
          surface: '#ffffff',
          inverseText: '#000000',
          text: '#ffffff',
        },
        contrast: {
          background: '#000000',
          surface: '#000000',
          primary: '#eeeeee',
          primaryVariant: '#eeeeee',
          secondary: '#ffff00',
          secondaryVariant: '#ffff00',
          text: '#ffffff',
          inverseText: '#000000',
          error: '#ff0000',
          warning: '#ffff00',
          info: '#0000ff',
          success: '#00ff00',
        },
      },
    }))
  }

  return { install }
}
