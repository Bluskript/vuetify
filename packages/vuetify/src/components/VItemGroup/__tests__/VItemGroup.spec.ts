// Components
import { VItemGroup, VItem } from '../'

// Utilities
import { defineComponent, h } from 'vue'
import { mount } from '@vue/test-utils'
import * as framework from '@/framework'

const Mock = defineComponent({
  name: 'Test',
  props: {
    disabled: Boolean,
  },
  setup (props) {
    return () => h(VItem, { disabled: props.disabled }, {
      default: (props: any) => {
        return h('div', { class: ['item', props.activeClass], onClick: props.toggle }, ['foo'])
      },
    })
  },
})

describe('VItemGroup', () => {
  it('should warn if using multiple prop without an array value', () => {
    jest.spyOn(framework, 'useVuetify').mockImplementation(() => ({
      defaults: {
        global: {},
      },
    }))

    mount(VItemGroup, {
      props: {
        multiple: true,
        value: '',
      },
    })

    expect('v-item-group modelValue must be an array when using multiple prop').toHaveBeenTipped()
  })

  it('should use provided active class', async () => {
    jest.spyOn(framework, 'useVuetify').mockImplementation(() => ({
      defaults: {
        global: {},
      },
    }))

    const wrapper = mount(VItemGroup, {
      slots: {
        default: () => [h(Mock), h(Mock)],
      },
    })

    await wrapper.find('.item').trigger('click')

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should select the first item if mandatory and no value', async () => {
    jest.spyOn(framework, 'useVuetify').mockImplementation(() => ({
      defaults: {
        global: {},
      },
    }))

    const wrapper = mount(VItemGroup, {
      props: { mandatory: true },
      slots: {
        default: () => h(Mock),
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  // https://github.com/vuetifyjs/vuetify/issues/5384
  // Not sure how to do this in 3.0, there does not seem to be
  // isDestroyed on vm? At least not typed through getCurrentInstance
  it.skip('should not unregister children when is destroyed', async () => {
    const Test = defineComponent({
      name: 'Test',
      props: { show: Boolean },
      setup (props) {
        return () => props.show ? h(VItemGroup, {
          modelValue: 0,
        }, () => h(Mock)) : h('div', ['something else'])
      },
    })
    const wrapper = mount(Test, {
      props: {
        show: true,
      },
    })

    const itemGroup = wrapper.findComponent(VItemGroup)

    await wrapper.setProps({ show: false })

    expect(itemGroup.emitted('update:modelValue')).toHaveLength(0)
  })

  // https://github.com/vuetifyjs/vuetify/issues/5000
  it('should update mandatory to first non-disabled item', async () => {
    jest.spyOn(framework, 'useVuetify').mockImplementation(() => ({
      defaults: {
        global: {},
      },
    }))

    const wrapper = mount(VItemGroup, {
      propsData: {
        mandatory: true,
      },
      slots: {
        default: () => [
          h(Mock, { disabled: true }),
          h(Mock),
        ],
      },
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })
})
