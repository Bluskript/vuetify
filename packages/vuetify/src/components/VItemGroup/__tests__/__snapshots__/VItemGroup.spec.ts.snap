// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VItemGroup should not unregister children when is destroyed 1`] = `
<div class="v-item-group">
  <div class="item">
    foo
  </div>
</div>
`;

exports[`VItemGroup should not unregister children when is destroyed 2`] = `
<div>
  something else
</div>
`;

exports[`VItemGroup should select the first item if mandatory and no value 1`] = `
<div class="v-item-group">
  <div class="item">
    foo
  </div>
</div>
`;

exports[`VItemGroup should update mandatory to first non-disabled item 1`] = `
<div class="v-item-group">
  <div class="item">
    foo
  </div>
  <div class="item v-item--active">
    foo
  </div>
</div>
`;

exports[`VItemGroup should use provided active class 1`] = `
<div class="v-item-group">
  <div class="item v-item--active">
    foo
  </div>
  <div class="item">
    foo
  </div>
</div>
`;
